SECTION A – CONCEPTUAL UNDERSTANDING
A1. Why are NumPy arrays preferred over Python lists?
Definition

NumPy array is a data structure provided by NumPy for storing and processing numerical data efficiently.

Answer

NumPy arrays are generally preferred over Python lists for numerical computation because they are faster and more memory-efficient. NumPy supports vectorized operations, which allow calculations on entire arrays without writing explicit Python loops. NumPy arrays also provide many mathematical, statistical, and linear-algebra functions that are useful in AI/ML.

Example

In [2]:
import numpy as np

a = np.array([1, 2, 3])
print(a * 2)

[2 4 6]


Key points

Faster numerical operations

Uses less memory

Supports vectorization

Provides mathematical functions

Very useful for AI/ML data processing

A2. Explain Vectorization
Definition

Vectorization is the process of performing an operation on an entire NumPy array at once instead of using an explicit Python loop.

Answer

Vectorization makes numerical computations faster because NumPy performs many operations internally using optimized code. It reduces the need for Python-level loops and becomes especially useful when working with large datasets.

In [1]:
#NORMAL LOOP
result = []
for x in [1, 2, 3]:
    result.append(x * 2)

In [2]:
#VECTORIZED NUMPY CODE
import numpy as np

a = np.array([1, 2, 3])
result = a * 2
print(result)

[2 4 6]


One-line explanation

Vectorization performs operations on the complete array efficiently without explicit Python loops.

A3. Difference between A * B and A @ B

Definition

A * B → element-wise multiplication

A @ B → matrix multiplication

In [3]:
#EXAMPLE
import numpy as np

A = np.array([[1, 2],
              [3, 4]])

B = np.array([[5, 6],
              [7, 8]])

print(A * B)
print(A @ B)

[[ 5 12]
 [21 32]]
[[19 22]
 [43 50]]


Explanation


A * B multiplies corresponding elements:


1×5 = 5

2×6 = 12

3×7 = 21

4×8 = 32


A @ B performs matrix multiplication.


Neural-network examples


A * B: Can be used for element-wise operations such as applying masks or element-wise activation-related calculations.


A @ B: Commonly used for calculating the weighted sum in a neural network:


X @ W + b

A4. What is Broadcasting?

Definition


Broadcasting is NumPy's ability to perform element-wise operations between arrays with different but compatible shapes.


General rule


When comparing shapes from right to left, dimensions are compatible when:


1.They are equal, OR

2.One of them is 1.

In [4]:
#EXAMPLE
import numpy as np

a = np.array([1, 2, 3])
b = 10

print(a + b)

[11 12 13]


One-line explanation

Broadcasting allows compatible arrays of different shapes to participate in element-wise operations.

A5. flatten() vs ravel()
Definition

Both flatten() and ravel() convert a multidimensional array into a one-dimensional array.

flatten()
Returns a copy of the array.

Changes to the result generally do not affect the original.

Uses additional memory.

ravel()
Returns a flattened view whenever possible.

Uses less memory.

May share memory with the original array.

In [5]:
#CODE
import numpy as np

a = np.array([[1, 2],
              [3, 4]])

print(a.flatten())
print(a.ravel())

[1 2 3 4]
[1 2 3 4]


Which is preferred for large images?

For a memory-constrained application, ravel() is generally preferred because it can avoid creating an unnecessary copy.

A6. Explain axis=0 and axis=1

Consider:

import numpy as np

a = np.array([[1, 2, 3],
              [4, 5, 6]])

Shape:

(2, 3)
axis=0

It performs the operation down the rows, producing one value for each column.

print(np.sum(a, axis=0))

Output:

[5 7 9]

Because:

1+4 = 5
2+5 = 7
3+6 = 9
axis=1

It performs the operation across the columns, producing one value for each row.

print(np.sum(a, axis=1))

Output:

[ 6 15]

Because:

1+2+3 = 6
4+5+6 = 15


A7. Normalization vs Standardization

Definition: Normalization

Normalization scales values into a fixed range, commonly 0 to 1.

Formula:

x' = (x - min) / (max - min)

Definition: Standardization

Standardization converts values so that they generally have:

Mean = 0
Standard deviation = 1

Formula:

z = (x - mean) / standard deviation

Why are they important?

Features may have very different scales. For example:

Age:       18–60
Salary:    20,000–500,000

Large-scale features may dominate certain ML algorithms.


In [6]:
#NumPy example
import numpy as np

x = np.array([10, 20, 30, 40, 50])

normalized = (x - x.min()) / (x.max() - x.min())
standardized = (x - x.mean()) / x.std()

print(normalized)
print(standardized)

[0.   0.25 0.5  0.75 1.  ]
[-1.41421356 -0.70710678  0.          0.70710678  1.41421356]


A8. Why does np.nanmean() exist?

Definition

np.nanmean() calculates the mean while ignoring NaN values.

NaN means Not a Number and represents missing or undefined numerical data.

In [7]:
#EXAMPLE
import numpy as np

data = np.array([10, 20, np.nan, 40])

print(np.mean(data))
print(np.nanmean(data))

nan
23.333333333333332


Explanation

np.mean() sees the NaN and returns nan.


np.nanmean() ignores NaN:


(10 + 20 + 40) / 3 = 23.33

A9. Shape of a batch of RGB images

Definition


In ML:


X.shape = (samples, features)


For RGB images, additional dimensions represent:


(batch, height, width, channels)


For:


64 images
Height = 128

Width = 128

RGB = 3 channels

Exact shape

(64, 128, 128, 3)

Explanation

64  → number of images

128 → height

128 → width

3   → RGB channels


A10. Why is reproducibility important?

Definition


Reproducibility means obtaining the same results when the same experiment is repeated under the same conditions.


Random numbers are frequently used in AI/ML for initialization, sampling, splitting data, etc.

In [8]:
#CODE
import numpy as np

rng1 = np.random.default_rng(42)
rng2 = np.random.default_rng(42)

print(np.array_equal(rng1.random(5), rng2.random(5)))

True


Explanation


Using the same seed produces the same sequence of random numbers, making experiments easier to reproduce and debug.

SECTION B – PREDICT THE OUTPUT


B1. Broadcasting

a = np.array([1, 2, 3])
b = np.array([[1], [2], [3]])

print((a + b).shape)

Shapes

a → (3,)
b → (3,1)

Broadcasting produces:

(3,3)

Output

(3, 3)

Explanation

The (3,) array is broadcast across the columns of the (3,1) array.

B2. Column indexing

x = np.arange(1, 7).reshape(2, 3)

print(x[:, 1])

First:

x =
[[1 2 3]
 [4 5 6]]


: means all rows.


1 means column index 1.


Output

[2 5]


B3. Boolean filtering

arr = np.array([5, 10, 15, 20, 25])

print(arr[arr % 10 == 0])

Values divisible by 10 are:

10, 20

Output

[10 20]

Explanation

Boolean indexing selects elements whose remainder after division by 10 is zero.


B4. Matrix multiplication


m = np.array([[1, 2],
              [3, 4]])

print(m.T @ m)

Transpose:

m.T =
[[1 3]
 [2 4]]

Multiplication:


[[1 3]     [[1 2]
 [2 4]]  @  [3 4]]

Output

[[10 14]
 [14 20]]


B5. mean() vs nanmean()


data = np.array([1, 2, np.nan, 4])

print(np.mean(data), np.nanmean(data))

Output

nan 2.3333333333333335

Explanation

np.mean() returns nan because the array contains NaN, while 

np.nanmean() ignores the NaN.


B6. np.where()


x = np.array([10, 20, 30])


y = np.where(x > 15, 1, 0)


print(y)


Check each value:


10 > 15 → 0

20 > 15 → 1

30 > 15 → 1


Output

[0 1 1]


B7. ndim and size

a = np.zeros((3, 4))


print(a.ndim, a.size)

Shape:

(3,4)

Number of dimensions:

2

Number of elements:

3 × 4 = 12

Output

2 12


B8. Reverse rows and columns

arr = np.arange(1, 13).reshape(3, 4)

print(arr[::-1, ::-1])

Original:

[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]]


Both axes are reversed.


Output

[[12 11 10  9]
 [ 8  7  6  5]
 [ 4  3  2  1]]

SECTION C – CODING / IMPLEMENTATION

The assignment specifically asks for vectorized NumPy code.


C1. Pass/Fail using np.where()


Definition

np.where(condition, value_if_true, value_if_false) selects values based on a condition.

In [10]:
import numpy as np

scores = np.array([35, 40, 55, 28, 75])

labels = np.where(scores >= 40, 'P', 'F')

print(labels)

['F' 'P' 'P' 'F' 'P']


Explanation

Scores greater than or equal to 40 are labelled P, otherwise F.

C2. Standardize a 1D array

Definition

Z-score standardization converts values using:


z = (x - mean) / standard deviation

In [11]:
import numpy as np

def standardize(x):
    return (x - np.mean(x)) / np.std(x)

x = np.array([10, 20, 30, 40, 50])

result = standardize(x)

print(result)

[-1.41421356 -0.70710678  0.          0.70710678  1.41421356]


Explanation

The function subtracts the mean and divides by the standard deviation.

C3. Combine two 2D arrays

Definition

np.vstack() vertically stacks arrays, meaning it combines them row-wise.

In [12]:
import numpy as np

batch1 = np.array([[1, 2],
                   [3, 4]])

batch2 = np.array([[5, 6],
                   [7, 8],
                   [9, 10]])

combined = np.vstack((batch1, batch2))

print(combined)
print(combined.shape)

[[ 1  2]
 [ 3  4]
 [ 5  6]
 [ 7  8]
 [ 9 10]]
(5, 2)


Explanation

vstack() combines the batches vertically because both arrays have the same number of columns.

C4. Column-wise mean, standard deviation and standardization

Definition

Column-wise standardization means standardizing each feature independently.

Given:

X.shape = (100,5)

there are:

100 samples
5 features

In [13]:
import numpy as np

X = np.random.default_rng(42).random((100, 5))

mean = X.mean(axis=0)
std = X.std(axis=0)

X_standardized = (X - mean) / std

print(X_standardized.shape)

(100, 5)


One-expression version

X_standardized = (X - X.mean(axis=0)) / X.std(axis=0)

Explanation

axis=0 calculates the mean and standard deviation separately for each feature column.

C5. Calculate y_pred and MSE

Definition

The linear model equation is:

y_pred = X @ w + b

MSE (Mean Squared Error) measures the average squared difference between actual and predicted values.

Formula:

MSE = mean((y_true - y_pred)²)

In [14]:
import numpy as np

X = np.array([[1, 2],
              [2, 3],
              [3, 4]])

w = np.array([0.5, 1.0])
b = 2.0

y_true = np.array([5, 7, 9])

y_pred = X @ w + b

mse = np.mean((y_true - y_pred) ** 2)

print("Predicted:", y_pred)
print("MSE:", mse)

Predicted: [4.5 6.  7.5]
MSE: 1.1666666666666667


Explanation

X @ w + b calculates predictions, and MSE calculates the average squared prediction error.

C6. Generate random samples with a fixed seed

Definition

np.random.default_rng(seed) creates a random number generator. A fixed seed makes the generated values reproducible.

In [15]:
import numpy as np

rng1 = np.random.default_rng(42)
data1 = rng1.normal(size=(10, 3))

rng2 = np.random.default_rng(42)
data2 = rng2.normal(size=(10, 3))

print(data1.shape)
print(np.array_equal(data1, data2))

(10, 3)
True


Explanation

The same seed produces identical random data.

C7. Min-max normalization of an image

Definition

Min-max normalization converts values to a specified range, here [0,1].

Formula:

(x - min) / (max - min)

In [16]:
import numpy as np

image = np.array([[0, 128],
                  [200, 255]])

normalized = (image - image.min()) / (image.max() - image.min())

print(normalized)

[[0.         0.50196078]
 [0.78431373 1.        ]]


C8. Average pixel value of each RGB image

Given:

shape = (4, 28, 28, 3)

Definition

np.mean() calculates the average value.
 To get one average value for each image, average across:


height → axis 1

width  → axis 2

channels → axis 3

In [17]:
import numpy as np

images = np.random.default_rng(42).random((4, 28, 28, 3))

average = images.mean(axis=(1, 2, 3))

print(average)
print(average.shape)

[0.49879444 0.49546077 0.48993754 0.50644914]
(4,)


Explanation

Axes (1, 2, 3) represent height, width, and RGB channels, leaving the image/batch axis unchanged.

Section D – Applied / Scenario-Based Problems

D1. Data Preprocessing Pipeline

You are given a raw 1D or 2D NumPy array containing NaN values.

Describe, step by step, the NumPy operations you would apply to:

(a) detect missing values,

(b) handle them appropriately and justify your strategy, and

(c) normalize or standardize the clean data before it is fed to a model.

Implement the full pipeline in code on a sample array of your own choosing, including at least one NaN.

Explanation

When a NumPy array contains NaN values, we can use the following steps:

Detect missing values using np.isnan().


Handle missing values using np.nanmean() to calculate the mean while ignoring NaN.

Replace the missing values with the calculated mean.

In [19]:
import numpy as np

data = np.array([10, 20, np.nan, 40, 50])

# Detect missing values
missing = np.isnan(data)
print("Missing values:", missing)

# Calculate mean ignoring NaN
mean_value = np.nanmean(data)

# Replace NaN with mean
clean_data = np.where(missing, mean_value, data)
print("Clean data:", clean_data)

# Standardize the clean data
standardized = (clean_data - np.mean(clean_data)) / np.std(clean_data)
print("Standardized data:", standardized)

Missing values: [False False  True False False]
Clean data: [10. 20. 30. 40. 50.]
Standardized data: [-1.41421356 -0.70710678  0.          0.70710678  1.41421356]


Explanation: np.isnan() detects missing values, np.nanmean() calculates the mean without considering NaN, and the missing value is replaced before standardization.

D2. Image Batch Understanding [8 marks]

Explain, with correct NumPy shapes, how a single RGB image differs from a batch of RGB images.

Write code to create a dummy batch of 16 images of size 64 × 64 using random pixel values.

Compute the average image of the batch, meaning the element-wise mean across all 16 images.

State the shape of your result and explain why.

Explanation

A single RGB image has the shape:

(64, 64, 3)

where:

64 → height

64 → width

3 → RGB color channels

A batch of 16 RGB images has the shape:

(16, 64, 64, 3)

Here, 16 represents the number of images.

In [20]:
import numpy as np

# Create a batch of 16 RGB images
images = np.random.randint(0, 256, size=(16, 64, 64, 3))

print("Batch shape:", images.shape)

# Compute average image across all 16 images
average_image = np.mean(images, axis=0)

print("Average image shape:", average_image.shape)

Batch shape: (16, 64, 64, 3)
Average image shape: (64, 64, 3)


Explanation:
 axis=0 averages the corresponding pixels across all 16 images while keeping the height, width, and RGB channels unchanged.

D3. Debug the Model Code 


The following code contains more than one bug:

X = np.array([[1, 2],
              [2, 3],
              [3, 4]])  # shape (3, 2)

w = np.array([0.5, 1.0, 0.2])  # intended weight vector

b = 2.0

y_pred = X @ w + b

y_true = np.array([10, 20, 30])

mse = np.mean(y_true - y_pred ** 2)

Identify every bug. Explain, in terms of shapes and operator precedence, why each is a problem, and rewrite the corrected code.

Bugs Identified


Bug 1 – Shape mismatch


X has shape (3, 2), so the weight vector must contain 2 values.


But:


w = np.array([0.5, 1.0, 0.2])


has shape (3,).


Therefore, X @ w cannot be performed.


The corrected weight vector is:


w = np.array([0.5, 1.0])


Bug 2 – Incorrect MSE calculation


The expression:


np.mean(y_true - y_pred ** 2)


does not calculate MSE correctly because exponentiation happens before subtraction.


The correct formula is:


np.mean((y_true - y_pred) ** 2)

In [21]:
#CORRECTED CODE
import numpy as np

X = np.array([[1, 2],
              [2, 3],
              [3, 4]])

w = np.array([0.5, 1.0])

b = 2.0

y_pred = X @ w + b

y_true = np.array([10, 20, 30])

mse = np.mean((y_true - y_pred) ** 2)

print("Predicted values:", y_pred)
print("MSE:", mse)

Predicted values: [4.5 6.  7.5]
MSE: 244.16666666666666


Explanation: 


The weight vector is changed to match the two features in X, and parentheses are used so that the prediction error is squared before calculating its mean.

Section E – Bonus / Challenge

E1. Per-Column Min–Max Normalization 

Question

Implement min-max normalization for a 2D feature matrix on a per-column (per-feature) basis, rather than using a single global min/max for the whole array. Your solution must be fully vectorized, with no explicit loops over rows or columns.


Answer


Min-max normalization scales each feature independently to the range 0 to 1. For each column, we subtract its minimum value and divide by the difference between its maximum and minimum values.

In [22]:
import numpy as np

X = np.array([[10, 100],
              [20, 200],
              [30, 300],
              [40, 400]])

column_min = X.min(axis=0)
column_max = X.max(axis=0)

X_normalized = (X - column_min) / (column_max - column_min)

print("Original matrix:")
print(X)

print("Column-wise normalized matrix:")
print(X_normalized)

Original matrix:
[[ 10 100]
 [ 20 200]
 [ 30 300]
 [ 40 400]]
Column-wise normalized matrix:
[[0.         0.        ]
 [0.33333333 0.33333333]
 [0.66666667 0.66666667]
 [1.         1.        ]]


Explanation



axis=0 calculates the minimum and maximum separately for each column. Broadcasting then applies these values to every row, so each feature is normalized independently without using any explicit loops.

E2. Three-Level Score Classification 

Question

Given a 1D array of exam scores, use a single vectorized expression (no if/else statements and no explicit loops) to classify each score as:


Fail (< 40)

Average (40–70 inclusive)

Excellent (> 70)


Hint: You may need to nest or combine np.where() calls.


Answer



We can use nested np.where() to classify all scores at once. The first condition checks whether the score is below 40. If not, the second np.where() checks whether the score is 70 or below.

In [23]:
import numpy as np

scores = np.array([25, 40, 55, 70, 85, 95])

result = np.where(scores < 40, "Fail",
                  np.where(scores <= 70, "Average", "Excellent"))

print("Scores:", scores)
print("Classification:", result)

Scores: [25 40 55 70 85 95]
Classification: ['Fail' 'Average' 'Average' 'Average' 'Excellent' 'Excellent']


Explanation


The first np.where() assigns Fail to scores below 40. For the remaining scores, the nested np.where() assigns Average to scores from 40 to 70 and Excellent to scores above 70. This performs the complete classification without loops or if/else statements.